src/Datadog.NativeLoader/proxy.cpp

```diff
@@ -0,0 +1,394 @@
+                std::string idValue = lineArray[0];
+                std::string osArchValue = lineArray[1];
+                std::string filepathValue = lineArray[2];
+
+                if (osArchValue == currentOsArch)
+                {
+                    // Convert possible relative paths to absolute paths using the configuration file folder as base
+                    // (current_path)
+                    filepathValue = std::filesystem::absolute(filepathValue).string();
+                    if (std::filesystem::exists(filepathValue))
+                    {
+                        guidBoolMap[idValue] = true;
+                        std::unique_ptr<DynamicInstance> instance =
+                            std::make_unique<DynamicInstance>(filepathValue, idValue);
+                        this->Add(instance);
+                        WSTRING env_key = WStr("PROFID_") + ToWSTRING(idValue);
+                        WSTRING env_value = ToWSTRING(filepathValue);
+                        bool envVal = SetEnvironmentValue(env_key, env_value);
+                        Debug("SetEnvVal: ", envVal, "; ", env_key, "=", env_value);
+                    }
+                    else if (guidBoolMap.find(idValue) == guidBoolMap.end())
+                    {
+                        guidBoolMap[idValue] = false;
+                    }
+                }
+            }
+        }
+        t.close();
+
+        // Set the current path to the original one
+        std::filesystem::current_path(oldCurrentPath);
+
+        for (const auto item : guidBoolMap)
+        {
+            if (!item.second)
+            {
+                Warn("Dynamic library for '", item.first, "' cannot be loaded");
+            }
+        }
+    }
+
+    HRESULT DynamicDispatcher::LoadClassFactory(REFIID riid)
+    {
+        HRESULT result = S_OK;
+        for (const auto& dynIns : m_instances)
+        {
+            if (dynIns != nullptr)
+            {
+                HRESULT localResult = dynIns->LoadClassFactory(riid);
+                if (FAILED(localResult))
+                {
+                    result = localResult;
+                }
+            }
+        }
+        return result;
+    }
+
+    HRESULT DynamicDispatcher::LoadInstance(IUnknown* pUnkOuter, REFIID riid)
+    {
+        HRESULT result = S_OK;
+        for (const auto& dynIns : m_instances)
+        {
+            if (dynIns != nullptr)
+            {
+                HRESULT localResult = dynIns->LoadInstance(pUnkOuter, riid);
+                if (FAILED(localResult))
+                {
+                    result = localResult;
+                }
+            }
+        }
+        return result;
+    }
+
+    HRESULT STDMETHODCALLTYPE DynamicDispatcher::DllCanUnloadNow()
+    {
+        HRESULT result = S_OK;
+        for (const auto& dynIns : m_instances)
+        {
+            if (dynIns != nullptr)
+            {
+                HRESULT localResult = dynIns->DllCanUnloadNow();
```

Comment:
According this https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-dllcanunloadnow
The possible values are `S_OK` and `S_FALSE` but we are handling `E_*` too.

So I changed the implementation to: 
```
    HRESULT STDMETHODCALLTYPE DynamicDispatcher::DllCanUnloadNow()
    {
        HRESULT result = S_OK;
        for (const std::unique_ptr<DynamicInstance>& dynIns : m_instances)
        {
            HRESULT localResult = dynIns->DllCanUnloadNow();
            if (FAILED(localResult))
            {
                Warn("Error calling DllCanUnloadNow in: ", dynIns->GetFilePath());
                result = localResult;
            }
            else if (localResult != S_OK)
            {
                // If we get something different than S_OK then we keep that result because the DLL cannot be unloaded.
                result = localResult;
            }
        }
        return result;
    }
```

in 6f12971

Reply:
Good point about the docs and thanks for the change. This way it's more robust.

✔️